File: Project_A/srcFilesForStudents_ProjectA/copy.c
// copy.c - Copy sourceFile to destinationFile

#include <stdio.h>

#define EOF -1

int main(int argc, char* args[]) {
    FILE* srcFile = NULL;
    FILE* dstFile = NULL;
    char* srcFilename = "<srcFilename>";
    char* dstFilename = "<dstFilename>";

    // Check the number of arguments.
    // In C, the command name 'copy' is considered as an argument.
    if (argc != 3) {
        printf("Usage: copy.exe <src> <dst> (on Windows)\n");
        return 0;
    }

    // Check if arguments are valid, if the srcFile exists, and if can create the dstFile.
    if (args[1] != NULL) { // Check <src>
        srcFilename = args[1];
        printf("copy: srcFilename = '%s'\n", srcFilename);
        srcFile = fopen(srcFilename, "r");
        if (srcFile == NULL) {
            printf("copy: Cannot open srcFile '%s'\n", srcFilename);
            return 0;
        }
    } else {
        printf("copy: [OK] srcFilename = '%s'\n", srcFilename);
    }

    if (args[2] != NULL) { // Check <dst>
        dstFilename = args[2];
        printf("copy: dstFilename = '%s'\n", dstFilename);
        dstFile = fopen(dstFilename, "w");
        if (dstFile == NULL) {
            printf("copy: Cannot open dstFile '%s'\n", dstFilename);
            return 0;
        }
    } else {
        printf("copy: [OK] dstFilename = '%s'\n", dstFilename);
    }

    // Execute the copy.
    int c;

    while ((c = fgetc(srcFile)) != EOF) {
        fputc(c, dstFile);
        putchar('.');
    }

    // Close and flush all the files.
    fclose(srcFile);
    fclose(dstFile);

    printf("copy: done.\n");
    return 1;
}

File: Project_A/src/Copy.java
// Copy.java - Copy sourceFile to destinationFile

import java.io.*;

public class Copy {
    private static int EOF = -1;
    private static File srcFile = null;
    private static File dstFile = null;
    private static String srcFilename = "<srcFilename>";
    private static String dstFilename = "<dstFilename>";

    private static void helper() {

        System.out.println("\n");
        System.out.println("CharCount:\t" + "Usage: java CharCount.java <src>");
        System.out.println("Copy:\t\t" + "Usage: java Copy.java <src> <dst>");
        System.out.println("LineCount:\t" + "Usage: java LineCount.java <src>");
        System.out.println("WordCount:\t" + "Usage: java WordCount.java <src>");
        System.out.println("WC:\t\t" + "Usage: java WC.java <src> { <src> }");

        System.out.println("\n");

    }

    public static void main(final String args[]) throws IOException {

        // Check the number of arguments.
        // In Java, the command name 'copy' is considered as an argument.

        boolean isHelper = args[0].equals("-?") || args[0].equals("-h") || args[0].equals("-help");

        if (isHelper) {
            helper();
            return;
        }

        if (args.length != 2) {
            System.out.println("Usage: copy <src> <dst>");
            return;
        }

        // Check if arguments are valid, if the srcFile exists, and if can create the
        // dstFile.
        if (args[0] != null) { // Check <src>
            srcFilename = args[0];
            System.out.println("copy: srcFilename = '" + srcFilename + "'");
            srcFile = new File(srcFilename);
            if (!srcFile.canRead()) {
                System.out.println("copy: Cannot open srcFile '" + srcFilename + "'");
                return;
            }
        } else {
            System.out.println("copy: [OK] srcFilename = '" + srcFilename + "'");
        }

        if (args[1] != null) { // Check <dst>
            dstFilename = args[1];
            dstFile = new File(dstFilename);
        } else {
            System.out.println("copy: [OK] dstFilename = '" + dstFilename + "'");
        }

        final FileInputStream srcStream = new FileInputStream(srcFile);
        final FileOutputStream dstStream = new FileOutputStream(dstFile);

        // Execute the copy.
        int c;

        while ((c = srcStream.read()) != EOF) {
            dstStream.write(c);
            System.out.print('.');
        }

        // Close and flush all the files.
        srcStream.close();
        dstStream.close();

        System.out.println("copy: done.");
    }
}
